src/sentry/utils/appleconnect/itunes_connect.py

```diff
@@ -3,349 +3,515 @@
-        if details_response.status_code == HTTPStatus.OK:
-            try:
-                data = details_response.json()
-                dsym_url: Optional[str] = safe.get_path(data, "data", "dsymurl")
-                return dsym_url
-            except Exception as e:
-                logger.info(
-                    "Could not obtain dSYM info for "
-                    "app id=%s, bundle_short=%s, bundle=%s, platform=%s",
-                    app_id,
-                    bundle_short_version,
-                    bundle_version,
-                    platform,
-                    exc_info=True,
-                )
-                raise e
-        return None
+    def request_sms_auth(self) -> None:
+        """Requests sending the authentication code to a trusted phone.
+
+        :raises ITunesError: if there was an error requesting to use the trusted phone.
+        """
+        assert self.state is ClientState.AUTH_REQUESTED
+        self._request_trusted_phone_info()
+        assert self._trusted_phone is not None
+        url = "https://idmsa.apple.com/appleauth/auth/verify/phone"
+        logger.debug("PUT %s", url)
+        response = self.session.put(
+            url,
+            json={
+                "phoneNumber": {"id": self._trusted_phone.id},
+                "mode": self._trusted_phone.push_mode,
+            },
+            headers={
+                "scnt": self._scnt,
+                "X-Apple-Id-Session-Id": self.session_id,
+                "Accept": "application/json",
+                "X-Apple-Widget-Key": self.service_key,
+                "Content-Type": "application/json",
+            },
+        )
+        if response.status_code != HTTPStatus.OK:
+            raise ITunesError("Unexpected response status: {response.status_code}")
+        self.state = ClientState.SMS_AUTH_REQUESTED
+
+    def sms_code(self, code: str) -> None:
+        """Sends the SMS auth code, completing authentication.
+
+        :raises InvalidSmsAuthError:
+        """
+        assert self.state is ClientState.SMS_AUTH_REQUESTED
+        assert self._trusted_phone is not None
+        url = "https://idmsa.apple.com/appleauth/auth/verify/phone/securitycode"
+        logger.debug("PUT {url}")
+        response = self.session.post(
+            url,
+            json={
+                "securityCode": {"code": code},
+                "phoneNumber": {"id": self._trusted_phone.id},
+                "mode": self._trusted_phone.push_mode,
+            },
+            headers={
+                "scnt": self._scnt,
+                "X-Apple-Id-Session-Id": self.session_id,
+                "Accept": "application/json",
+                "X-Apple-Widget-Key": self.service_key,
+            },
+        )
+        if response.status_code != HTTPStatus.OK:
+            # TODO: Make invalid code distinguishable from generic error.
+            raise InvalidAuthCodeError
+        else:
+            self.state = ClientState.AUTHENTICATED
```

Comment:
should this be setting `self._trusted_phone = None` once this step is done? it doesn't look like this value gets preserved when a client is exported in a non `ClientState.SMS_AUTH_REQUESTED` state, which suggests to me that this shouldn't be populated once the client transitions out of `ClientState.SMS_AUTH_REQUESTED`.

Reply:
Kind of similar answer as above, maybe but it doesn't make much of a difference.   Still think the separate authenticator could do this nicely as it could go back to NEW and reset everything once you have a client (you can't move like in rust)